### ISSUE TITLE
Code Contribution: 【Hard】【Operator Development】grid_sampler_2d & grid_sampler_2d_backward

### ISSUE THREAD

**Issue by kiddyjinjin** [FlagOpen/FlagGems#1117]
## Description 任务介绍

Develop forward and backward functions for 2D grid sampler operators.  
开发 2D 版本的 grid sampler 算子前向与反向功能。

涉及两个算子：
- `grid_sampler_2d`
- `grid_sampler_2d_backward`

## Requirements 任务要求

__Interface 接口__

Forward:
`grid_sampler_2d(Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners) -> Tensor`

Backward:
`grid_sampler_2d_backward(Tensor grad_output, Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool[2] output_mask) -> (Tensor, Tensor)`

__Function reference 功能参考__

- PyTorch functional API: `torch.nn.functional.grid_sample`  
  https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html
- Chinese mirror / explanation:  
  https://docs.pytorch.ac.cn/docs/stable/generated/torch.nn.functional.grid_sample.html

__Implementation reference 实现参考__

- ATen native GridSampler implementation (algorithm and behavior):  
  https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/GridSampler.cpp
- FlagGems upsample example (Triton kernel & dispatch style reference):  
  https://github.com/FlagOpen/FlagGems/blob/master/src/flag_gems/ops/upsample_nearest2d.py

__Behavior 行为要求__

- The operators should support all arguments defined in the interfaces above:
  - `interpolation_mode`
  - `padding_mode`
  - `align_corners`
  - `output_mask` (in backward)
  
  算子应完整支持接口中定义的所有参数：
  - `interpolation_mode`
  - `padding_mode`
  - `align_corners`
  - 反向中的 `output_mask`（用于控制是否对 input / grid 计算梯度）。

- Semantics must be fully aligned with `torch.nn.functional.grid_sample` for 2D inputs:
  - `input` shape: `(N, C, H_in, W_in)`
  - `grid` shape: `(N, H_out, W_out, 2)`
  - `grid` values in normalized coordinates `[-1, 1]` (左上角 `(-1, -1)`，右下角 `(1, 1)`）
  - Interpolation modes and padding modes behavior must match PyTorch.

  算子语义需与 `torch.nn.functional.grid_sample` 在 2D 情况下保持一致，包括：
  - `input` 形状 `(N, C, H_in, W_in)`
  - `grid` 形状 `(N, H_out, W_out, 2)`
  - `grid` 为归一化坐标，范围在 `[-1, 1]`
  - 插值模式 / padding 模式均需与 PyTorch 对齐。

- Backward:
  - `grid_sampler_2d_backward` should compute gradients w.r.t. `input` and `grid` consistent with PyTorch.
  - `output_mask[0]` 控制是否计算 `∂L/∂input`，`output_mask[1]` 控制是否计算 `∂L/∂grid`，当对应 mask 为 `False` 时允许跳过该部分梯度计算以提升性能。

  反向传播实现需与 PyTorch 对应算子在数值和行为上保持一致。

- Determinism behavior should be consistent with PyTorch (it is acceptable that backward is non-deterministic on some backends, same as upstream implementation).

  关于确定性（determinism）行为与 PyTorch 保持一致即可；在部分后端上反向可能是非确定性的。

## DDL 提交时间

Please submit a Pull Request within 2 weeks after accepting the assignment.  
请于接取任务后两周内提交 PR。

Please provide both accuracy test and performance test code.  
请同时提供实现正确性测试与性能测试代码。

**Comment by bin913** [comment]
The latest repo already contains an implementation for this operator
